### ISSUE TITLE
Add a PIDFile class that can be used as a context

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8659]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)| @wsanchez reported|
|-|-|
|Trac ID|trac#8659|
|Type|enhancement|
|Created|2016-07-23 03:47:08Z|
|Branch|https://github.com/twisted/twisted/tree/8659-pidfile|

Add a PIDFile class that can be used as a context:

```
with PIDFile(path) as pidfile:
   ...
```

Intended client is runner; may be more broadly useful.

<details><summary>Searchable metadata</summary>

```
trac-id__8659 8659
type__enhancement enhancement
reporter__wsanchez wsanchez
priority__normal normal
milestone__None None
branch__8659_pidfile 8659-pidfile
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__twist twist
time__1469245628708603 1469245628708603
changetime__1483659992359193 1483659992359193
version__None None
owner__None None

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_3"></a>|@wsanchez commented|
|-|-|

Putting up for review.

`documentation` builder is complaining about:
```
twisted.application.runner._runner.RunnerOptions.pidFile: invalid ref to 'NamedConstant' resolved as 'constantly.NamedConstant'
```

There are actually many of these, but only this one is "new" due to an unrelated change to that `@cvar` docstring.

This seems erroneous; `NamedConstant` is a defined symbol in that module's namespace.  I've filed https://github.com/twisted/pydoctor/issues/128 for that.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez removed owner<br>@wsanchez set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

Opened a PR here: https://github.com/twisted/twisted/pull/572

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez set owner to @wsanchez|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)<a name="note_7"></a>|@moshez commented|
|-|-|

(Review available in PR, removed keyword and assigned back to wsanchez)

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_8"></a>|@wsanchez commented|
|-|-|

OK I addressed everything in the review, I think.

There's enough churn here that another review is probably a good idea, so back in the queue.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph commented|
|-|-|

The "runner" component refers to `twisted.runner`, which is a package for running _other_ processes.  This is `twisted.application.runner`, and the intended client is `twist`, both of which are part of the core.  So, switching the component back to core.

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_13"></a>|@markrwilliams commented|
|-|-|

https://github.com/twisted/twisted/pull/572#pullrequestreview-5768232

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set owner to @wsanchez<br>@wsanchez set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_15"></a>|@wsanchez set status to `closed`|
|-|-|

Merged.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph set status to `reopened`|
|-|-|

Reverted in e05482c4b97f60afb22985611b5788bffb83b231.  See https://github.com/twisted/twisted/pull/613

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_17"></a>|@wsanchez commented|
|-|-|

New PR: https://github.com/twisted/twisted/pull/620

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez removed owner<br>@wsanchez set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_20"></a>|@glyph set status to `closed`|
|-|-|

This was fixed in 29599b1c409e4401fd39b8d6d1d7a4f5e03b6fc8 but the commit message was misformatted - for future reference, https://twistedmatrix.com/trac/wiki/ReviewProcess#Authors:Howtomergethechangetotrunk describes the format .  (Technically speaking it wasn't reviewed first either, but I can see based on the discussion that implicitly it looked like a full review had taken place, and it had been approved previously so I'm going to let it slide...)
